### ISSUE TITLE
Wrong typoscript in flux configuration

### ISSUE THREAD

**Issue by cmichael-de** [FluidTYPO3/flux#1671]
**EDIT by Claus: please read https://github.com/FluidTYPO3/flux/issues/1671#issuecomment-467828171 for cause and reason for saying "won't fix"**

I define a typoscript constant and call it inside the setup.txt settings in my provider extension. Now I can use the constant value in the flux template in the `<f:section name="Configuration">` like this `<flux:field.select  name="class" items="{settings.cssClasses}"/>`. So far it is working fine!

I create a second root page for another website in the same typo3 cms. I set another value for {settings.cssClasses} in the constant editor. Now the trouble is starting! The value for `{settings.cssClasses}` is every time the value that I defined in the first root page.

When I swap the root page positions the value changes.

This only happens inside the  `<f:section name="Configuration">`. I tested and used `{settings.cssClasses}` inside the  `<f:section name="Main">` in the frontend I got the correct value from current root page.

**Comment by rupasix** [comment]
Related: https://github.com/FluidTYPO3/flux/issues/1453

**Comment by NamelessCoder** [comment]
*TL;DR: Extbase features are semi-broken in backend context and Flux is affected.*

This response is going to be somewhat controversial, so I am hoping for your understanding:

This is an unfortunate combination of Extbase and more traditional TS resolving not playing along nicely. To be specific, this is caused by ConfigurationManager 1) not receiving or resolving the correct page UID, and therefore 2) not resolving the correct TS.

The logic that selects the TS template will then look for:

* A TS template record marked as "root"
* Limiting selection to a single record
* But not ordering by any specific properties

This means that _you get the first root TS record that was inserted into the site_.

Furthermore, this is complicated by the fact that when you read a data structure from a Flux form, this happens _without any page context_ because it has to happen at a point before a page UID can be (safely) resolved.

The short version is that you should never use page-context-specific variables in your `Configuration` section because it may be read in contexts that simply do not have a page UID anywhere.

History
-------

There was a time when Flux contained overrides for ConfigurationManager to add at least a minimum of additional lookup strategies (most importantly: resolving PID from record being edited). This meant that it was impossible to use a custom ConfigurationManager and this decision was unmade, in favor of simply using whatever resolving the TYPO3 core provides - semi-broken or not - and then advising people to push for this being fixed in TYPO3.

Because this override historically made more trouble than benefit, and requires maintaining a class that is essentially a copy of a TYPO3 core class with some modifications, it was dropped and will not return.

In essence: Extbase is created to work with backend contexts but unlike frontend contexts, resolving a page UID in backend is a pretty complicated task because there is no uniform method used in TYPO3 to, for example, always put a page UID in the URL in a given parameter.

What can we do about this?
----------------------------

This is where the controversial parts come in. *We will no longer attempt to solve this issue through overrides* - instead, I will now refer you to the TYPO3 core bug tracker to open/find/reanimate issues relating to incorrect resolving of page UID in Extbase contexts.

Then:

* Suggest two strategies: adapt ConfigurationManager and make a universal rule about including page UID in backend requests in a predictable way, including when you edit a record, perform a copy, translate and so on.
* Proceed with temporarily fixing ConfigurationManager to at least cover the one case described here; when you edit a record.

It is possible to demonstrate this problem with any number of integrations that trigger when you, for example, edit a record. You can try this out - if you create a user function that fills items in a `select` TCA component by using `itemsProcFunc` and in this user function you try to use ConfigurationManager to resolve TS, you will experience this very problem - without any connection to Flux.

In semi-strong words: we will no longer try to fix TYPO3 core bugs with overrides. In this specific case, doing it 1) introduces differences not just in Flux but literally every other extension you have installed and 2) obligates us to maintain this class forever.

Workarounds
-------------

All is not lost if you wish to achieve your use case, but none of the workarounds are truly pretty.

1. Use a ViewHelper that will read TS in exactly the way you need it to (for example, reading only your specific extension scope) then assign the resolved variables and render the ViewHelper. Put the Flux form definition inside this. Make sure the VH knows what to do in case no UID can be resolved.
2. Avoid caching the DS generated by Flux. This is the default - but if you used the `static` Flux form option make sure you disable that.
3. Use hooks in Flux that will trigger when a form is resolved, then manipulate that (same strategy as the VH but doesn't require changing the template). You can find the relevant hook and instructions on https://github.com/FluidTYPO3/flux/commit/1f76e8a60b99d7b93ee8752345573530d4b81d8e#diff-c83ba51878737bacf58618479e99b55aR17.

I do apologise for saying plain old *no* to "fixing" this but hopefully you understand. The symptom you experience has nothing whatsoever to do with Flux _except_ for being in a core class that Flux uses - but this type of problem (and some very hairy TCA ones...) is far more likely to be encountered when you use Flux.

PS: #1453 is very, very loosely related but differes from this particular problem in that it pertains to pageTSconfig that's generated as one huge piece containing TS conditions. As such it is a separate issue that doesn't have to do with ConfigurationManager and that issue will not be affected at all by solving this one - and vice versa.

**Comment by NamelessCoder** [comment]
@rupasix I also hope this info might help you avoid #1453 cases or track down the specific reason why your cached pageTSconfig contains two identical blocks... it is perhaps traceable to using page-context-specific variables in the `Configuration` section?

**Comment by rupasix** [comment]
We gave up to solve this issue. We've handled that project somehow so in the end it was not a problem. I think i understand your statement and i agree with you about not modyfing core logic on that level, broken or not. As always, thanks for detailed explanation - now i get it that issue is actually not related :)

**Comment by NamelessCoder** [comment]
Closing as "won't fix" - best we can do is hope/encourage that core solves the issue somehow.
